### PR TITLE
Add an XRoot storage backend

### PR DESCRIPTION
**Pull request by jhiemstrawisc**
Some non-pelican origins currently in the OSDF don't speak HTTP. This makes them incompatible with Pelican and is preventing us from fully transitioning legacy services. To work around that limitation, this adds an "xroot" backend, which allows us to stand a Pelican origin (speaking HTTP) in front of an xroot-only origin.

This backend isn't meant to be extensible, and as such I haven't added any tests. To test by hand, you can configure:
```
Origin:
  StorageType: xroot
  EnablePublicReads: true
  XRootServiceUrl: <some xroot-only origin url, with port>
  FederationPrefix: <a valid prefix for that origin>
  StoragePrefix: <same as federation prefix>
```

### CODE REVIEW

**Comment by jhiemstrawisc** [comment]
I'll also note that `server_utils/origin.go` needs a lot of cleanup at this point. I intentionally avoided that in this PR because I wanted to minimize any risk when we backport this to 7.8. I've opened another issue to handle this and some other cleanup in issues #1286 and #1287.

**Review by haoming29 — CHANGES_REQUESTED** [review]
A couple of minor changes requested, plus the followings:

* When running the origin with `xroot` backend, the self-tests are failing:
```
WARNING[2024-05-15T15:57:58Z] Self-test monitoring cycle failed:  Test file transfer failed during upload: Failed to start request for test file upload: Put "https://6ee28c5df997:8443/pelican/monitoring/self-test-2024-05-15T15:57:48Z.txt": net/http: timeout awaiting response headers 
```

We should stop the self-test if it is expected to fail, or should note this in the log. I'd prefer stopping it to avoid spamming the log with the error messages every 15s

* The test origin/file didn't respond: `root://fndca1.fnal.gov:1095//pnfs/fnal.gov/usr/des/persistent/stash/testfile` when I was testing locally. We should find another available file to test before merging this in.

File: docs/parameters.yaml
Comment (jhiemstrawisc) [reply docs/parameters.yaml:861]: I know this xroot backend is meant to be a one-off to facilitate transitioning legacy OSDF caches to their Pelicanized counterparts, and so it's not necessarily extensible. That being said, I think you're right that half documenting it isn't the way to go about that. I'll make the updates :)

**Comment by haoming29** [comment]
So I did the following manual tests. Most of them passed. Some of them didn't:
* Run an origin and pull the file through `curl` origin endpoint: ✅ 
* Run a local federation in pelican mode, pull the file through using pelican client: ✅ 
* Run a local federation in osdf mode, set direct read on the origin, pull the file using pelican client: ✅ 
* Run a local federation in osdf mode, only enable read and public read at the origin, pull the file using pelican client: 🚫 
  * If the director makes it to return my local pelican cache, it works
  * if the director didn't find my local cache and used other caches in topology it fails with the following error messages:
 
```console
ERROR[2024-05-16T17:16:33Z] Failure transferring /pnfs/fnal.gov/usr/des/persistent/stash/testfile: Attempt #3: failed download from dtn-pas.denv.nrp.internet2.edu:8443: server returned 500 Internal Server Error (5.3s elapsed, 27.6s since start); Attempt #2: failed download from osdf-uw-cache.svc.osg-htc.org:8443: server returned 500 Internal Server Error (5.1s elapsed, 22.2s since start); Attempt #1: failed download from dtn-pas.jack.nrp.internet2.edu:8000: failed connection setup: Get "http://dtn-pas.jack.nrp.internet2.edu:8000/pnfs/fnal.gov/usr/des/persistent/stash/testfile": net/http: timeout awaiting response headers (17s since start) 
```

**Comment by jhiemstrawisc** [comment]
I think that single failure is okay and expected. When you're running the local federation, OSDF caches won't be able to discover the xroot backend because they're still pointed at the actual OSDF director (which your local xroot backend isn't advertising to). That being said, it's somewhat concerning that they respond with an HTTP 500... That sounds like a separate issue though

**Comment by haoming29** [comment]
> I think that single failure is okay and expected. When you're running the local federation, OSDF caches won't be able to discover the xroot backend because they're still pointed at the actual OSDF director (which your local xroot backend isn't advertising to). That being said, it's somewhat concerning that they respond with an HTTP 500... That sounds like a separate issue though

How would OSDF legacy caches know about this origin? I thought they only know origins from the topology, and it seems like `/pnfs` prefix is already in the topology, so even if this origin is up in the production environment, would those legacy cache unaware of that?

**Review by haoming29 — APPROVED** [review]

**Comment by jhiemstrawisc** [comment]
I think the idea is that we'd have to replace the existing `/pnfs` namespace with the new origin because the old origin doesn't speak HTTP. Attempts 2 and 3 are hitting Pelican caches, which also forward their requests as HTTP and not xroot, so that might actually explain the 500s.

**Comment by jhiemstrawisc** [comment]
@bbockelm I'll leave this one open through tomorrow in case you have time for a review. Otherwise I'll merge and have @turetske patch into 7.8 to unblock transitioning caches.

**Pull request merged**
